activesupport/lib/active_support/cache/key.rb

```diff
@@ -0,0 +1,160 @@
+    #   k1 = Key.new("foo")
+    #   k1 << "stay-puft"
+    #   k2 = Key.new(["foo", "bar"])
+    #   k1.cache_key == k2.cache_key
+    #   # => true
+    #   k3 = Key.new(["foo", ["bar"]])
+    #   k2.cache_key == k3.cache_key
+    #   # => true
+    #
+    # It is faster to directly add objects to the
+    # Key object via `<<` than to generate
+    # an intermediate array of objects
+    #
+    # The class supports strings, arrays, hashes,
+    # objects with a `cache_key` method and anything
+    # that can be implicitly cast to a string:
+    #
+    # Examples:
+    #
+    #   Key.new("foo").cache_key
+    #   # => "foo"
+    #
+    #   Key.new(["foo", "bar"]).cache_key
+    #   # => "foo/bar"
+    #
+    #   Key.new({ bar: 1, foo: 2 }).cache_key
+    #   # => "bar=1/foo=2"
+    #
+    #   klass = Class.new { def cache_key; "foo"; end }
+    #   Key.new(klass.new).cache_key
+    #   # => "foo"
+    #
+    # In addition to generating a `cache_key`
+    # a `cache_version` is also generated if the object
+    # passed in responds to `cache_version` or if it contains
+    # an element that responds to `cache_version`.
+    #
+    # The purpose of keeping this value seperate from
+    # `cache_key` is to support the use of "cache versioning"
+    # where the same key is recycled but the version of the object
+    # being stored is kept directly in the cache.
+    #
+    # Example:
+    #
+    #   klass = Class.new { def cache_version; "foo"; end }
+    #   Key.new(klass.new).cache_version
+    #   # => "foo"
+    #
+    #   klass = Class.new { def cache_version; "foo"; end }
+    #   Key.new([klass.new]).cache_version
+    #   # => "foo"
+    #
+    #   klass = Class.new { def cache_version; "foo"; end }
+    #   Key.new({foo: klass.new}).cache_version
+    #   # => "foo"
+    class Key # :nodoc:
+      attr_reader :cache_key, :cache_version
+
+      def self.to_key(key)
+        return key if key.is_a?(Key)
+
+        self.new(key)
+      end
+
+      def initialize(key = nil)
+        @cache_key = +""
+        @cache_version = +""
+        @empty_key = true
+        self << key
+      end
+
+      def update(key)
+        if key.respond_to?(:cache_key)
+          store_key(key.cache_key.to_s)
+          store_version(key)
+          return
+        end
+
+        case key
+        when Array
+          key.each do |k|
+            self << k
+          end
+        when Hash
+          key = key.to_a
+          store_version(key)
+          key.sort_by! { |k, _| k.to_s }
+          key.collect! { |k, v| "#{k}=#{v}" }
+
+          key.each do |k|
+            store_key(k)
+          end
+        else
+          store_version(key)
+          store_key(key)
+        end
+      end
+      alias :<< :update
+
+      private
+        def store_key(key)
+          @cache_key << "/" unless @empty_key
+          @empty_version =
+          @cache_key << key.to_param.to_s
+          @empty_key = false
+        end
+
+        def store_version(key)
```

Comment:
This method feels like it's handling a lot of different things at once - dealing with saving cache versions from individual keys as well as looping through array keys in order to recursively handle the cache version of each key. I wonder if it makes sense to handle those two responsibilities separately?

Reply:
I split `store_key` out to it's own method as it was called multiple places from `update`. I like that this logic for actually adding the key is inside the same method while all other conditions in the case are recursive, it helps me to distinguish between the two.

Reply:
Hmm, lemme illustrate what I'm imagining in my head and if you like it better the way it is now then it's fine - more of a SRP issue than anything technical.

```ruby
def store_version(key)
  split_object(key) unless key.respond_to(:cache_version)
  @cache_version << "/" unless @cache_version.empty?
  @cache_version << key.cache_version.to_param.to_s
end

def split_object(key)
  if key.is_a?(Array)
    key.each { |k| store_version(k) }
  elsif key.respond_to?(:to_a) 
    split_object(key.to_a)
  end
end
```